### PR TITLE
(WIP) Standard bounties integration

### PR DESCRIPTION
**Pull request by owocki**
# TODO List

https://github.com/ConsenSys/StandardBounties/

## Development
* [X] posting a bounty
* [ ] claiming (in standard bounties parlace, fulfilling) a bounty
* [ ] approving (in standard bounties parlace, accepting) a bounty claim
* [ ] clawing back expired funds (after a bounty expires) (in standard bounty parlace, kill a bounty)

## Backend
* [ ] backend bounty sync
* [ ] update database model (if needed)

## Test
* [ ] Test end to end with ETH token
* [ ] Test end to end with multiple fundings
* [ ] Test end to end with ERC20 token
* [ ] Test what bounties.network bounties look like when ingested to gitcoin
* [ ] Test visa versa

### CODE REVIEW

**Comment by owocki** [comment]
@rajatkapoor do you still want this

**Comment by owocki** [comment]
cc @jhaas -- im rejecting the current claim on this since i have not heard from the claimer

**Comment by jasonrhaas** [comment]
ok sounds good.  I can point on this one, I might break it up into several tickets/PRs once I scope out the work a bit more.

**Comment by abitrolly** [comment]
What is this "standard bounties"?

**Comment by owocki** [comment]
@abitrolly https://github.com/ConsenSys/StandardBounties/

**Comment by jasonrhaas** [comment]
FYI I am working off a branch in my forked repo over here -- https://github.com/jasonrhaas/web/tree/standard_bounties_integration

Mostly its just comments and stuff like at this point.  I'll start to sync it over to this PR once it makes sense.

**Comment by jasonrhaas** [comment]
Alright so I had a good call with Mark Beylin about how the Standard Bounties system works.  Here is my summary and the plan forward for getting this into Gitcoin:

### Creating new Bounties - Gitcoin Way

The current way of funding new Gitcoin Bounties is via the Gitcoin Smart Contract.  A new Gitcoin Smart Contract needs to be created on the network of choice (Mainnet, Rinkeby, etc).  Once that is created, a new bounty can be created.  For example, I created a new bounty on the Ropsten network, and that added a new transaction from my account to the Gitcoin Smart Contract on the Ropsten network.  Each new bounty that is created adds a transaction to the specific contract that is sitting on the network.

For example:
https://ropsten.etherscan.io/address/0x577ec61dB11b49286839f502481776F0Be01104D

### Creating new Bounties - Standard Bounties Way

Standard Bounties uses IPFS to manage the bounties on the network.  Standard Bounties has contracts deployed on the mainnet and rinkeby network.  The contract itself is always looking for new data that is pushed to IPFS.  

In practice, a JSON object with the bounty information is created, and then sent to IPFS.  IPFS then returns a hash which is how the data is accessed within the contract.  The contract that lives on the network is always looking for changes to this location in IPFS (need to check on this), and it will automatically append the bounty into to that contract once it sees it.

### Plan for Integrating Standard Bounties (Test using Rinkeby network)

1.  **Create new bounty** from Gitcoin, and then issue + activate it on the bounties network.  If successful, this should show up in the BountiesFactory explorer.
2.  **Fulfill bounty**.  Standard Bounties has the ability for multiple people to submit work for a bounty.  This functionality could be implemented in the Gitcoin interface or we can just use the bounties network.
3.   **Accept fulfillment**.  The arbiter for the contract can be the created or someone else that has been specified.
4.  **Bonus Round** .  It would be nice if we could code into the bounty the ability to pay out the bounty automatically once the Github Pull Request is merged in.

File: app/assets/v2/js/abi.js [hunk 1]
```diff
@@ -1,23 +1,23 @@
...
         case "kovan":
             throw "not supported";
         break;
         case "rinkeby":
-            throw "not supported";
+            throw "0xd5ea74ad5bcec859ea7a7104401ea1c746c84074";
         break;
         case "custom network":
-            return '0x0ed0c2a859e9e576cdff840c51d29b6f8a405bdd';
+            return '0xbe716ab501a10d9ea0362bf20a9b77aba1bd2d88';
```
Comment (jasonrhaas) [inline app/assets/v2/js/abi.js]: What is the custom network in terms of Standard Bounties?  From what I understand only mainnet and rinkeby is supported right now.

File: app/economy/eth.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 def get_network_details(network):
     bounty_contract_address = None
     if network == 'mainnet':
-        bounty_contract_address = '0x4e1799e6e54d848394b0b90593e789e939a9f801'
+        bounty_contract_address = '0xb10700b5ece20a3c65b047f76fd3dc13720bd30e'
```
Comment (jasonrhaas) [inline app/economy/eth.py]: Should this be the same address as above?  In the `abi.js` file the mainnet address used is `0x066128b9f7557b5398db3d4ed141f2e64245ffa1`.

**Comment by jasonrhaas** [comment]
Is there a way to combine my forked working branch with what is in this PR?  I can create a new PR by attempting to merge into master, or can attempt to merge in my branch into this existing branch, but I believe that would create a new PR as well.

**Comment by owocki** [comment]
> Is there a way to combine my forked working branch with what is in this PR? I can create a new PR by attempting to merge into master, or can attempt to merge in my branch into this existing branch, but I believe that would create a new PR as well.

maybe you could just branch off of the branch in the WIP PR?

i'm happy to go whatever way you want.

**Comment by owocki** [comment]
> Plan for Integrating Standard Bounties (Test using Rinkeby network)


This looks good to me.

File: app/assets/v2/js/abi.js (hunk 1 above)
Comment (owocki) [reply app/assets/v2/js/abi.js]: custom network is `testrpc`

testrpc just a way to create a localhost ethereum network => https://www.npmjs.com/package/ethereumjs-testrpc

highly HIGHLY recommend this approach for testing and quick iteration.. ephemeral `testrpc` nodes are cheap and very easy to develop upon

File: app/economy/eth.py (hunk 2 above)
Comment (owocki) [reply app/economy/eth.py]: i... think so.  i may have just gotten this wrong (or i wrote this PR when there was a diff version of `StandardBoutnties` deployed to the mainnet.

File: app/assets/v2/js/abi.js (hunk 1 above)
Comment (jasonrhaas) [reply app/assets/v2/js/abi.js]: Is that address the always the same for a custom network though?  It would vary based on the test environment right?

Comment (owocki) [reply app/assets/v2/js/abi.js]: depends on how you start testrpc.

i use this script , https://github.com/gitcoinco/smart_contracts/blob/master/scripts/testrpc.bash , which uses the same seed every single launch.  thus saving me the extra step of having to update this everytime i deploy a change

**Comment by vs77bb** [comment]
@owocki @jasonrhaas Kevin mentioned to me this is progressing nicely... just checking in to make sure. Nice to meet you, Jason!

**Comment by jasonrhaas** [comment]
hey @vs77bb, nice to meet you!  I've updated https://github.com/gitcoinco/web/pull/140 with detailed instructions on how to test the standard bounties integration.  Even if you don't want to fire up the code, there are screenshots so you can get a gist of what is going on.

**Comment by vs77bb** [comment]
@jasonrhaas Awesome work!

**Comment by vs77bb** [comment]
@owocki this one still shows in 'Work Submitted'... my guess is this is complete / needs pay out?

**Comment by owocki** [comment]
doing that now..

**Comment by jasonrhaas** [comment]
Did this one pay out?  I see its expired but I didn't see any notifications or anything

**Comment by owocki** [comment]
talked to jason offline.. we handled the payment offchain.

**Pull request closed without merging**